Fix writing of buffered data, ensure chunks fit within 64kb SCTP packets

nathanoehlman opened the pull request:
This does a number of things:

- In the event that `_write` is called with a chunk that exceeds 64kb in size, that chunk is further broken down into smaller chunks that do fit within that size. This prevents failures on the data channel due to attempting to write more data to the data channel than it could handle in a single chunk (https://bugs.chromium.org/p/webrtc/issues/detail?id=2270). 

- Fixes the resuming of sending buffered data once the channel has recovered, in particular, the situation where there was more than 1 queued chunk. Due to the checking for the existence of queued chunks in `_write`, this meant that buffered writes were never processed, but just left waiting. Buffered writes will now write, and will also correctly respond to changes in the channel conditions (readyState or bufferedAmount)

DamonOehlman commented:
Nice work Nathan - LGTM.  It might be worth seeing if you can make use of [`stream-spec`](https://github.com/dominictarr/stream-spec) which should give you some additional confidence that the second case has been covered from a testing perspective also.

nathanoehlman commented:
Cool - thanks @DamonOehlman. That `stream-spec` module looks interesting as well.

The other thing I neglected to mention in this PR was that I removed the exception handler that swallowed up the `NetworkError` when attempting to write after a close (but before the `close` event had been fired). I'm in two minds about this - I've removed it because the error that I got from attempting to write too much data to the channel is the same exception as if you attempt to write to the channel after it's been closed. In the latter case, it's fine to swallow it up, but in the former, not so much (need to restart the data channel in most cases) - so it seems like that exception is a bit more general.

Now that the chunking is in, there _shouldn't_ be any more exceptions thrown as a result of attempting to write too much data, so I'm tempted to put it back in - because the write after close exceptions are incredibly annoying.

On the other hand, seeing as how generic it is, there's the chance that more valid exceptions will be swallowed up instead.

Thoughts?

silviapfeiffer commented:
That's quite the patch!

Would be good if the app didn't fail in case of an exception, but did maybe put a warning in the console?
